### ISSUE TITLE
Add variable dropout probability within `__call__`?

### ISSUE THREAD

**Issue by mathDR** [patrick-kidger/equinox#1015]
Hi!  I was wondering what people's thoughts were regarding allowing for runtime dropout probability assignment?

For instance the current `__call__` method would be changed to something like:
```
    @named_scope("eqx.nn.Dropout")
    def __call__(
        self,
        x: Array,
        *,
        p: Float | None = None,
        key: PRNGKeyArray | None = None,
        inference: bool | None = None,
        deterministic: bool | None = None,
    ) -> Array:
        """**Arguments:**

        - `x`: An any-dimensional JAX array to dropout.
        - `p`: The dropout probability, if None, we use `self.p`
        - `key`: A `jax.random.PRNGKey` used to provide randomness for calculating
            which elements to dropout. (Keyword only argument.)
        - `inference`: As per [`equinox.nn.Dropout.__init__`][]. If `True` or
            `False` then it will take priority over `self.inference`. If `None`
            then the value from `self.inference` will be used.
        - `deterministic`: Deprecated alternative to `inference`.
        """

        if deterministic is not None:
            inference = deterministic
            warnings.warn(
                "Dropout()(deterministic=...) is deprecated "
                "in favour of Dropout()(inference=...)"
            )
        if p is None:
            p = self.p
        if inference is None:
            inference = self.inference
        if isinstance(self.p, (int, float)) and self.p == 0:
            inference = True
        if inference:
            return x
        elif key is None:
            raise RuntimeError(
                "Dropout requires a key when running in non-deterministic mode."
            )
        else:
            q = 1 - lax.stop_gradient(p)
            mask = jrandom.bernoulli(key, q, x.shape)
            return jnp.where(mask, x / q, 0)
```

Thoughts?

**Comment by patrick-kidger** [comment]
I think this is probably too niche to be included in Equinox itself -- but you can define an analogous module to use in your own code! :)
